styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -215,12 +186,101 @@ private void ensureSecrets(WorkflowInstance workflowInstance, RunSpec runSpec) {
         LOG.info("[AUDIT] Workflow {} refers to secret {}",
                  workflowInstance.workflowId(), specSecret.name());
       }
+    });
+  }
+
+  private void ensureServiceAccountKeySecret(WorkflowInstance workflowInstance, RunSpec runSpec) throws IOException {
+    if (!runSpec.serviceAccount().isPresent()) {
+      return;
+    }
+
+    final String serviceAccount = runSpec.serviceAccount().get();
+
+    // Check that the service account exists
+    final boolean serviceAccountExists = serviceAccountKeyManager.serviceAccountExists(serviceAccount);
+    if (!serviceAccountExists) {
+      LOG.error("[AUDIT] Workflow {} refers to non-existent service account {}", workflowInstance.workflowId(),
+          serviceAccount);
+      throw new InvalidExecutionException("Referenced service account " + serviceAccount + " was not found");
+    }
+
+    final String secretName = buildSecretName(serviceAccount);
+
+    LOG.info("[AUDIT] Workflow {} refers to secret {} of {}", workflowInstance.workflowId(), secretName,
+        serviceAccount);
+
+    // TODO: shard locking to regain concurrency
+    synchronized (secretMutationLock) {
+
+      // Check if we have a valid service account key secret already
+      final Secret existingSecret = client.secrets().withName(secretName).get();
+      if (existingSecret != null) {
+
+        if (serviceAccountKeysExist(existingSecret)) {
+          return;
+        }
+
+        LOG.warn("[AUDIT] Service account keys have been deleted for {}, recreating",
+            serviceAccount);
+
+        // Need to delete this secret before creating a new one
+        client.secrets().delete(existingSecret);
+      }
+
+      // Create service account keys and secret
+      final ServiceAccountKey jsonKey;
+      final ServiceAccountKey p12Key;
+      try {
+        jsonKey = serviceAccountKeyManager.createJsonKey(serviceAccount);
+        p12Key = serviceAccountKeyManager.createP12Key(serviceAccount);
+      } catch (IOException e) {
+        LOG.error("[AUDIT] Failed to create keys for {}", serviceAccount, e);
+        throw e;
+      }
+
+      final Map<String, String> keys = ImmutableMap.of(
+          STYX_WORKFLOW_SA_JSON_KEY, jsonKey.getPrivateKeyData(),
+          STYX_WORKFLOW_SA_P12_KEY, p12Key.getPrivateKeyData()
+      );
+
+      final Map<String, String> annotations = ImmutableMap.of(
+          STYX_WORKFLOW_SA_JSON_KEY_NAME_ANNOTATION, jsonKey.getName(),
+          STYX_WORKFLOW_SA_P12_KEY_NAME_ANNOTATION, p12Key.getName(),
+          STYX_WORKFLOW_SA_SECRET_ANNOTATION, serviceAccount
+      );
+
+      final Secret newSecret = new SecretBuilder()
+          .withNewMetadata()
+          .withName(secretName)
+          .withAnnotations(annotations)
+          .endMetadata()
+          .withData(keys)
+          .build();
+
+      client.secrets().create(newSecret);
+
+      LOG.info("[AUDIT] Secret {} created for {} referred to by workflow {}",
```

Comment:
"For" might read ambiguously, as in "for [use by this service account]", while it's actually "storing [this service account's private key!!!11]".